### PR TITLE
Solid harm prefact

### PR DESCRIPTION
**Pull request by arthur-lin1027**
The inner product obtaining our expansion coefficients erroneously contained the solid harmonic prefactor. I added code to divide through by the prefactor. This causes the expansion coefficients to actually converge to the underlying atomic gaussian, which is necessary. 

I currently cleaning up an example notebook that shows that the expansion of an isotropic gaussian does indeed converge.

However, an anisotropic gaussian does not yet converge... I have some suspicions.

### CODE REVIEW

File: tests/test_convergence.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        Test that coefficients are correct, such that the approximation using these coefficients in the expansion
+        can reasonably recreate the original atomic gaussian (in this case, an unnormalized gaussian with sigma=1)
+        """
+
+        frame = Atoms(positions=np.array([[0, 0, 0]]), numbers=[0])
+        frame.arrays["quaternions"] = np.array([[1, 0, 0, 0]])
+
+        frame.arrays[r"c_diameter[1]"] = 2 * sigma * np.ones(1)
+        frame.arrays[r"c_diameter[2]"] = 2 * sigma * np.ones(1)
```
Comment (SeonwooH) [inline tests/test_convergence.py:25]: where is c_diameter []? Why are you also multiplying by np.one(). Wouldn't it just be the same value? Why is this necessary?

File: anisoap/utils/moment_generator.py [hunk 2]
```diff
@@ -129,8 +129,11 @@ def compute_moments_inefficient_implementation(A, a, maxdeg):
         <x^{n_0} * y^{n_1} * z^{n_2}> = \int(x^{n_0} * y^{n_1} * z^{n_2}) * \exp(-0.5*(r-a).T@cov@(r-a)) dxdydz
         \sum_{i=1}^{\\infty} x_{i}
 
-        Note that the term "moments" in probability theory are defined for normalized Gaussian distributions.
-        Here, we take the Gaussian
+        Note that the term "moments" in probability theory are defined for normalized Gaussian distributions. These
+        recursive relations find the moments of a normalized Gaussian, but we actually want to find the moments of
+        the unnormalized underlying gaussian (as seen in the equation above) because calculating expansion
```
Comment (SeonwooH) [inline anisoap/utils/moment_generator.py:134]: Does knowing moments of unnormalized Gaussian expansions help determine the coefficients? How?

File: tests/test_convergence.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        frame.arrays[r"c_diameter[3]"] = 2 * sigma * np.ones(1)
+        frames = [frame]
+
+        # An rgw that's a little bigger than the atom gaussian sigma converges well and is sufficient for this test.
+        rgw = sigma + 0.5
+
+        max_radials = range(10)
+        errs = []
+        for max_radial in max_radials:
+            HYPER_PARAMETERS = {
```
Comment (SeonwooH) [inline tests/test_convergence.py:35]: What are hyper parameters

File: tests/test_convergence.py (hunk 1 above)
Comment (arthur-lin1027) [reply tests/test_convergence.py:25]: c_diameter[1], c_diameter[2], c_diameter[3] are the dimensions of the ellipsoid.

Multiplying by np.ones(1) looks weird, but I'm multiplying by np.ones(n) because each of the c_diameter fields needs to be length n, where n is the number of ellipsoids in the frame. Here, I'm just testing convergence on a single particle, but later, I may want to test convergence on multiple particles.

File: tests/test_convergence.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                "basis_rcond": 1e-14,
+                "basis_tol": 1e-1,
+            }
+            representation = EllipsoidalDensityProjection(**HYPER_PARAMETERS)
+            descriptor_raw = representation.transform(frames, normalize=True)
+            descriptor = metatensor.operations.sum_over_samples(
+                descriptor_raw, sample_names="center"
+            )
+
+            def evaluate_bases(r):
```
Comment (SeonwooH) [inline tests/test_convergence.py:52]: I do not understand what is this function for and what it's returning.

File: tests/test_convergence.py (hunk 3 above)
Comment (arthur-lin1027) [reply tests/test_convergence.py:35]: Hyperparameters are parameters that you choose to run your model with.
This is opposed to parameters, which is what your model tries to find to best fit your data.

e.g. if you're trying to fit a polynomial to a bunch of points, like
$y=a_0x^0 + a_1x^1 + ... + a_nx^n$, 

n is a hyperparameter that you choose beforehand, to determine what order polynomial you want to fit to your data (n=1 would be a linear fit, n=2 would be a quadratic fit, etc.). After you choose your hyperparameter n, the model determines $a_0, a_1, ...$, which are the parameters that best fit the data.

File: anisoap/utils/moment_generator.py (hunk 2 above)
Comment (arthur-lin1027) [reply anisoap/utils/moment_generator.py:134]: Yes, this is necessary. We do want to find the moments of the unnormalized gaussian, since we're taking the moments of an effective gaussian that isn't unnormalized.

File: tests/test_convergence.py (hunk 4 above)
Comment (arthur-lin1027) [reply tests/test_convergence.py:52]: This function outputs what the basis function $Y_{lm}*R_{nl}$ is, evaluated at the point $r$

File: anisoap/utils/moment_generator.py (hunk 2 above)
Comment (SeonwooH) [reply anisoap/utils/moment_generator.py:134]: So our goal is to find the moments of unnormalized guassian from normalized ones? Just to clarify, we are finding unnormalized guassian moments because they are the actual representations coarse-grained system?

File: pyproject.toml [hunk 5]
```diff
@@ -14,12 +14,10 @@ commands =
 
 [tool.coverage.run]
 branch = true
-data_file = 'tests/.coverage'
+data_file = "tests/.coverage"
```
Comment (SeonwooH) [inline pyproject.toml:17]: I am just wondering if is there a specific reason for changing it from single to double quotation mark?

File: tests/test_convergence.py (hunk 3 above)
Comment (SeonwooH) [reply tests/test_convergence.py:35]: So we can choose the number of hyperparameters to determine the coefficient of bases? Would deciding what parameters affect the computational cost and accuracy of prediction?

File: tests/test_convergence.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                        return np.real(sph_harm(m, l, theta, phi))
+                    else:
+                        return (
+                            np.sqrt(2)
+                            * (-1.0) ** m
+                            * np.real(sph_harm(m, l, theta, phi))
+                        )
+
+                bases = descriptor.copy()
+                for key, block in bases.items():
```
Comment (SeonwooH) [inline tests/test_convergence.py:71]: this seems like transforming from bases that work for isotropic particles to those for anisotropic... am I right? What is this loop exactly doing?

File: anisoap/representations/ellipsoidal_density_projection.py [hunk 7]
```diff
@@ -129,7 +133,11 @@ def pairwise_ellip_expansion(
                         deg = l + 2 * (num_ns[l] - 1)
                         moments_l = moments[: deg + 1, : deg + 1, : deg + 1]
                         values_ldict[l].append(
-                            np.einsum("mnpqr, pqr->mn", sph_to_cart[l], moments_l)
+                            np.einsum(
+                                "mnpqr, pqr->mn",
+                                sph_to_cart[l] / solid_harm_prefact[l],
+                                moments_l,
+                            )
```
Comment (SeonwooH) [inline anisoap/representations/ellipsoidal_density_projection.py:140]: Are we adding the matrix-matrix multiplication between the coefficients and moments? Why do we compute and add this?

File: tests/test_convergence.py (hunk 6 above)
Comment (arthur-lin1027) [reply tests/test_convergence.py:71]: Not really. When I do descriptor.copy(), my intention is to copy all the metadata that's in the descriptor into a new bases `TensorMap`. 

`descriptor` is a `TensorMap` that contains the **coefficients** evaluated at every `n,l,m` index. These `n,l,m` indices **are the metadata** that we want to retain in our new `TensorMap` `bases`. The intention of `bases` is to contain all the evaluations $R_{nl}*Y_{lm}$ at the same `n,l,m` that was in descriptor.

To see if the coefficients can successfully recreate the underlying gaussian, I multiply all the coefficients by all the evaluations, which I do later.

File: anisoap/representations/ellipsoidal_density_projection.py (hunk 7 above)
Comment (arthur-lin1027) [reply anisoap/representations/ellipsoidal_density_projection.py]: This piece of code combines the moments and the cartesian representation of the spherical harmonics to obtain each coefficient that is present in descriptor. It is replicating equation A30 (in the appendix) in the [AniSOAP paper](https://arxiv.org/abs/2403.19039)

**Review by SeonwooH — APPROVED** [review]
looks great!

File: anisoap/utils/moment_generator.py (hunk 2 above)
Comment (arthur-lin1027) [reply anisoap/utils/moment_generator.py:134]: The unnormalized gaussian moments aren't the actual representation, but is a major component.

Oversimplified, but if $c$ is our actual representation, and $c = a * b * ...$, the moments can be thought of as the "$a$" component.

File: tests/test_convergence.py (hunk 3 above)
Comment (arthur-lin1027) [reply tests/test_convergence.py:35]: Absolutely. Choosing higher values for hyperparameters gives us higher-resolution descriptors (i.e more coefficients) but is more computationally expensive.

**Review by rosecers — APPROVED** [review]

**Comment by curiosity54** [comment]
hey, i had a rather quick look, but was the purpose here to modify the density expansion coefficients? I am afraid that adding the additional 1/sqrt{2l+1} affects how higher order features are constructed down the line

**Comment by arthur-lin1027** [comment]
> hey, i had a rather quick look, but was the purpose here to modify the density expansion coefficients? I am afraid that adding the additional 1/sqrt{2l+1} affects how higher order features are constructed down the line could you perhaps give me an overview of how the density expansion is now normalized?

Hi Jigyasa,

I made these changes under the assumption that a linear combination of expansion coefficients and basis functions should recreate the underlying gaussian. I identified that this 1/sqrt(2l+1) factor was erroneously included and needed to be divided out in order to recreate the underlying gaussian.

**Comment by arthur-lin1027** [comment]
To reiterate one more time, I'm trying to unify equations 15 and 16 within the [AniSOAP paper](https://arxiv.org/pdf/2403.19039), but right below equation 16, we make an incorrect statement saying that 

> with $R_l^m(r) = r^l * Y_l^m(\hat{r})$

This is incorrect, because actually $R_l^m(r) = \sqrt{\frac{4\pi}{2l+1}}r^l * Y_l^m(\hat{r})$. Hence I need to divide out this factor, because in the code, we are calculating $R_l^m$, but we actually want $r^l * Y_l^m(\hat{r})$ (eqn 15 in the paper)

**Comment by curiosity54** [comment]
okay, thank you! i was a little unsure of the prefactor being there in the definition of real-valued spherical harmonics but all good if this makes the anisoap and isotropic gaussian expansion consistent. (incidentally, this made me realize there's a typo in Eq 17, l' must be equal to l)

File: anisoap/representations/ellipsoidal_density_projection.py [hunk 8]
```diff
@@ -79,6 +79,13 @@ def pairwise_ellip_expansion(
     keys = [tuple(i) + (l,) for i in keys for l in range(lmax + 1)]
     num_ns = radial_basis.get_num_radial_functions()
     maxdeg = np.max(np.arange(lmax + 1) + 2 * np.array(num_ns))
+    # This prefactor is the solid harmonics prefactor, that we need to divide by later.
+    # This is needed because spherical_to_cartesian calculates solid harmonics Rlm = sqrt((4pi)/(2l+1)) * r^l*Ylm
+    # Our expansion coefficients from the inner product does not have this prefactor included, so we divide it later.
+    solid_harm_prefact = np.sqrt((4 * np.pi) / (np.arange(lmax + 1) * 2 + 1))
+    scaled_sph_to_cart = []
+    for l in range(lmax + 1):
+        scaled_sph_to_cart.append(sph_to_cart[l] / solid_harm_prefact[l])
```
[the remark this answers is not in the document]
Comment (arthur-lin1027) [reply anisoap/representations/ellipsoidal_density_projection.py:89]: cannot actually make this change, because sph_to_cart is a list of 5-dimensional arrays that change depending on l. `np.divide` cannot understand how to unify the dimensions so I just use a for loop, plus it's more explicit here.

**Comment by rosecers** [comment]
That’s fine, only a suggestion
